File: python/signature.py
# -*- coding: utf-8 -*-
import base64
from hashlib import sha1


def get_signature(secret_key, params):
    u"""
    >>> params = {
    ...     "merchant": 43210,
    ...     "amount": '174.7',
    ...     "currency": 'RUB',
    ...     "description": u'Заказ №73138754',
    ...     "order_id": '73138754',
    ...     "success_url": 'http://myshop.ru/success/',
    ...     "fail_url": 'http://myshop.ru/fail/',
    ...     "cancel_url": 'http://myshop.ru/cart/',
    ...     "signature": '',
    ...     "unix_timestamp": '1399461194',
    ...     "meta": '{"tracking": 1234}',
    ...     "salt": '00000000000000000000000000000000',
    ... }
    >>> get_signature('C0FFEE', params)
    'cb3743cc37d87f5a4255fc3a99c223c0e869c145'
    """
    return double_sha1(secret_key, '&'.join(
        force_str(k) + '=' + base64.b64encode(force_str(params[k]))
        for k in sorted(params)
        if params[k] and k != 'signature'
    ))


def force_str(v):
    return v.encode('utf-8') if isinstance(v, unicode) else str(v)


def double_sha1(secret_key, data):
    """
    >>> double_sha1('C0FFEE', 'example')
    '27d204596505ff298ca79fb3bb949501cd7b2fa7'
    """
    for i in range(2):
        data = sha1(secret_key + data).hexdigest()
    return data


if __name__ == "__main__":
    import doctest
    doctest.testmod()
    print 'doctests passed'

File: php/signature.php
<?php

function get_signature($secret_key, $params) {
    $keys = array_keys($params);
    sort($keys);
    $chunks = array();
    foreach ($keys as $k) {
        $v = (string) $params[$k];
        if (($v !== '') && ($k != 'signature')) {
            $chunks[] = $k . '=' . base64_encode($v);
        }
    }
    return double_sha1($secret_key, implode('&', $chunks));
}

function double_sha1($secret_key, $data) {
    return sha1($secret_key . sha1($secret_key . $data));
}

$secret_key = 'C0FFEE';
$params = array(
    "merchant" => 43210,
    "amount" => '174.7',
    "currency" => 'RUB',
    "description" => 'Заказ №73138754',
    "order_id" => '73138754',
    "success_url" => 'http://myshop.ru/success/',
    "fail_url" => 'http://myshop.ru/fail/',
    "cancel_url" => 'http://myshop.ru/cart/',
    "signature" => '',
    "unix_timestamp" => '1399461194',
    "meta" => '{"tracking": 1234}',
    "salt" => '00000000000000000000000000000000',
);

$signature = get_signature($secret_key, $params);
echo "signature = '$signature'\n";
assert($signature == '163807001c1997be21352ea53fa6e5f4e988483d');

assert(get_signature($secret_key, array('testing' => 'x')) == '22b6416d3d9ece6d969d40c62097f1b1878b0b89');
assert(get_signature($secret_key, array('testing' => '0')) == 'c869d507973a539f73e5c5fe08017001171cfd1e');
assert(get_signature($secret_key, array('testing' => 0)) == 'c869d507973a539f73e5c5fe08017001171cfd1e');
